klippy/extras/firmware_retraction.py

```diff
@@ -14,22 +14,46 @@ def __init__(self, config):
         self.unretract_speed = config.getfloat('unretract_speed', 10., minval=1)
         self.unretract_length = (self.retract_length
                                  + self.unretract_extra_length)
+        self.lift_z = config.getfloat('lift_z', 0., minval=0.)
+        self.current_lift_z = 0.
         self.is_retracted = False
+        self.prev_G0 = None
+        self.prev_G1 = None
         self.gcode = self.printer.lookup_object('gcode')
         self.gcode.register_command('SET_RETRACTION', self.cmd_SET_RETRACTION,
                                     desc=self.cmd_SET_RETRACTION_help)
         self.gcode.register_command('GET_RETRACTION', self.cmd_GET_RETRACTION,
                                     desc=self.cmd_GET_RETRACTION_help)
         self.gcode.register_command('G10', self.cmd_G10)
         self.gcode.register_command('G11', self.cmd_G11)
-
     def get_status(self, eventtime):
         return {
             "retract_length": self.retract_length,
             "retract_speed": self.retract_speed,
             "unretract_extra_length": self.unretract_extra_length,
             "unretract_speed": self.unretract_speed,
+            "lift_z": self.lift_z,
         }
+
+    def _override_gcode_move(self, undo=False):
+        if not undo and not self.prev_G0 and not self.prev_G1:
+            self.prev_G0 = self.gcode.register_command('G0', None)
+            self.gcode.register_command('G0', self.cmd_G0)
+            self.prev_G1 = self.gcode.register_command('G1', None)
+            self.gcode.register_command('G1', self.cmd_G1)
+        elif self.prev_G0 and self.prev_G1:
+            self.gcode.register_command('G0', None)
+            self.gcode.register_command('G0', self.prev_G0)
+            self.gcode.register_command('G1', None)
+            self.gcode.register_command('G1', self.prev_G1)
+            self.prev_G0 = self.prev_G1 = None
```

Comment:
This code looks a bit 'ugly' from far away, but I'm unable to come up with a nicer implementation. I'm interested in feedback. Also are people fine with the `func(undo=True/False)` pattern rather than `func()/func_reset()` for example?